lib/watcherjob.go

```diff
@@ -73,20 +104,92 @@ func (job *watcherJob) eventLoop(ctx context.Context) error {
 	logger.Get(ctx).Debug("Watcher connected")
 	job.SetReady(true)
 
-	process := MustGetProcess(ctx)
-
 	for {
 		select {
 		case event := <-watcher.Events():
-			process.Spawn(func(ctx context.Context) error {
-				return job.eventFunc(ctx, event)
-			})
+			if err := job.enqueue(ctx, event); err != nil {
+				logger.Get(ctx).WithError(err).Error("Failed to enqueue an event")
+			}
 		case <-watcher.Done():
 			return trace.Wrap(watcher.Error())
 		}
 	}
 }
 
+// enqueue puts an event to a temporary micro-queue allocated for each resource to maintain the event order.
+func (job *watcherJob) enqueue(ctx context.Context, event types.Event) error {
+	queueKey := watcherQueueKey{kind: event.Resource.GetKind(), name: event.Resource.GetName()}
+
+	var (
+		queue *watcherQueue
+		exist bool
+	)
+	for {
+		if err := ctx.Err(); err != nil {
+			return trace.Wrap(err)
+		}
+
+		job.queueMu.Lock()
+		if queue, exist = job.queues[queueKey]; !exist {
+			if len(job.queues) >= job.config.MaxConcurrency {
+				job.queueMu.Unlock()
+				continue
+			}
```

Comment:
> Checking for the desired condition in a tight loop like this is bad for performance. 

Yes that definitely must be improved.

> Side note: why do we even need separate queues? Looks to me like all events end up being proceeded by the same function anyhow. Why not just use a single large queue that buffers all events?

@fspmarshall 

That was a temporary workaround against the `UpdatePluginData()` race. My concern was the following:

- If i'm processing events sequentially, it removes the race (at least for the case when there's only one plugin process).
- But processing events sequentially is slower than processing them concurrently.
- So why not have a mixed approach:
  - Events for different access requests: process them concurrently.
  - Events for the same access request: process sequentially to avoid races.

I added tests for both cases explaining the approach.

`UpdatePluginData` races in 99.9% cases are running around the same request so why not have a treatment for this?

As I said, that workaround was *temporary* and now it seems not to be required BUT I tend to keep it because it prevents potential race conditions just not allowing them to happen at least in the context of one process.